File: 03-RECURSION/reverse-linked-list.py
from typing import *

# https://leetcode.com/problems/reverse-linked-list/
class Solution:
    def reverseList(self, head: ListNode):
        if not head or not head.next:
            return head

        new_head = self.reverseList(head.next)

        head.next.next = head
        head.next = None

        return new_head

File: 03-RECURSION/Go/reverse-linked-list.go
package main

// https://leetcode.com/problems/reverse-linked-list/

type ListNode struct {
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	newHead := reverseList(head.Next)

	head.Next.Next = head
	head.Next = nil

	return newHead
}
